### PR TITLE
Timelock safety checks

### PR DESCRIPTION
**Pull request by j-baker**
**Goals (and why)**: Timelock has some sharp edges. Let's dull them.

**Implementation Description (bullets)**:

1. Make it impossible to overflow. You can deadlock yourself, but that is fixable. Overflow is not.
2. Have users indicate whether this is a new instantiation of the service. It's not cool for timelock to create the data dir if it's not a new service. It won't deal with every case, but it seems like it'd provide a lot of real world mileage.

**Testing (What was existing testing like?  What have you done to improve it?)**:

This class is only used internally and so can only be tested internally. It should need a default config change, but nothing more.

**Concerns (what feedback would you like?)**:

Whether there are easier ways to achieve the same effect.

**Where should we start reviewing?**:

**Priority (whenever / two weeks / yesterday)**:

Probably worth getting this in pretty fast.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/3498)

### CODE REVIEW

File: timelock-agent/src/main/java/com/palantir/timelock/config/PaxosInstallConfiguration.java [hunk 1]
```diff
@@ -27,8 +27,31 @@ default File dataDirectory() {
...
+                            + "be set to false for safety reasons.");
+        }
+
+        if (!isNewService() && !dataDirectory().isDirectory()) {
+            throw new IllegalArgumentException(
+                    "If this is a new timelock service, please configure paxos.is-new-service to true for the first "
+                            + "startup only of each node. \n\nOtherwise, the timelock data directory appears to no "
+                            + "longer exist. If you are trying to move the nodes on your timelock cluster or add new "
+                            + "nodes, you have likely already made a mistake by this point. This is a non-trivial "
+                            + "operation, please be careful and do not be afraid to seek out help.");
```
Comment (tboam) [inline timelock-agent/src/main/java/com/palantir/timelock/config/PaxosInstallConfiguration.java]: I think we can be more direct here and instruct people to speak to the Atlas team before proceeding.

File: timelock-agent/src/main/java/com/palantir/timelock/config/PaxosInstallConfiguration.java [hunk 2]
```diff
@@ -27,8 +27,31 @@ default File dataDirectory() {
         return new File("var/data/paxos");
     }
 
+    /**
+     * Set to true if this is a new stack. Otherwise, set to false.
+     */
+    @JsonProperty("is-new-service")
+    boolean isNewService();
```
Comment (gokcanozakdag) [inline timelock-agent/src/main/java/com/palantir/timelock/config/PaxosInstallConfiguration.java]: what would be the default value here?

Comment (jeremyk-91) [reply timelock-agent/src/main/java/com/palantir/timelock/config/PaxosInstallConfiguration.java]: There is no default value, and we'll throw an ISE if we try and build an object without it.

This is probably reasonable - in internal distributions we can specify it explicitly (probably `false`), and users can override it as needed using standard internal protocols.

File: docs/source/release_notes/release-notes.rst [hunk 3]
```diff
@@ -50,6 +50,12 @@ develop
     *    - Type
          - Change
 
+    *    - |improved|
+         - A few timelock ops edge cases have been removed. Timelock users must now indicate whether they are booting their
+           servers for the first time or subsequent times, to avoid the situation where a timelock node becomes newly
+           misconfigured and thinks it is booting up for the first time again.
+           (`Pull Request <https://github.com/palantir/atlasdb/pull/3498>`__)
+
```
Comment (jeremyk-91) [inline docs/source/release_notes/release-notes.rst]: nit: The overflow thing is a win and maybe worth calling out (you could say that it's unlikely anyone actually hits this, to be fair, but I think it's still nice)

File: timelock-agent/src/main/java/com/palantir/timelock/config/PaxosInstallConfiguration.java (hunk 1 above)
Comment (j-baker) [reply timelock-agent/src/main/java/com/palantir/timelock/config/PaxosInstallConfiguration.java]: i've cited 'appropriate approvals' - in short these messages with a 'who' on the end have typically ended up fairly out of date...

File: timelock-agent/src/test/java/com/palantir/timelock/config/PaxosInstallConfigurationTest.java [hunk 4]
```diff
@@ -28,45 +27,65 @@
...
+    }
+
+    @Test
+    public void throwsIfConfiguredToBeExistingServiceWithoutDirectory() {
+        File mockFile = getMockFileWith(false, true);
+
+        assertThatThrownBy(ImmutablePaxosInstallConfiguration.builder()
+                .dataDirectory(mockFile)
+                .isNewService(false)
+                ::build).isInstanceOf(IllegalArgumentException.class);
```
Comment (jeremyk-91) [inline timelock-agent/src/test/java/com/palantir/timelock/config/PaxosInstallConfigurationTest.java]: nit: we can probably extract a method for the assertion?

**Pull request merged**